text/3778-proc-macro-in-same-package-as-app.md

```diff
@@ -0,0 +1,171 @@
+Currently, we create a new proc macro as so:
+1. Create a new package
+2. In its cargo.toml, specify that it is a proc macro package
+3. In the main project, add the package as a dependency
+4. Implement the proc macro in the new package
+
+After this change, we create a new proc macro like this:
+1. Implement the proc macro in a new `procmacros.rs` in `proc-macro`.
+
+To build only the macro, use:
+```console
+$ cargo build --macros
+```
+
+## Importing
+[importing]: #importing
+To use the proc macro, simply import it via `macros::*`.
+```rust
+use macros::my_macro;
+```
+
+Note that macros is only available to inside the package (i.e. bin, lib, examples...). This means that one would have to reexport the macros in `lib.rs` in order for users of a library to use it. It would still be available in `main.rs`, `tests`, `examples`, etc, though.
+
+## An example
+Suppose you are developing a library that would have normal functions as well as proc macros. The file structure would look like this:
+```
+My-Amazing-Library
+|---src
+|   |---lib.rs
+|   |---procmacros.rs
+|   |---common.rs
+|---cargo.toml
+```
+`common.rs` is a normal file that declares common data structures and functions. `procmacros.rs` defines macros, which will be made available to `lib.rs`. `lib.rs` can use the macros defined, and reexport the macros to make it available to anyone using the library.
+
+Using code in `common.rs` in `procmacros.rs` is like how you would normally:
+```rust
+mod common;
+use common::*;
+```
+
+Now, to make the macros available, reexport them, and if you want gate a macro behind a feature flag, it would be like how you would normally also, with cfg:
+```rust
+// in lib.rs
+#[cfg(feature = "my_feature")]
+pub use macros::a_niche_macro;
+pub use macros::a_common_macro; // (not gated)
+```
+
+Finally, testing is also how you would expect it. It would be made available to the `tests` directory. ([importing])
+
+## Cargo.toml configs
+Libraries like `syn`, `quote`, and `proc-macro2`, would be included under `[build-dependecies]` in the cargo.toml. (Perhaps we should put it in a new dependency section for proc macros?)
+
+Like `tests` or `lib`, this would have its own `[proc-macro]` section in cargo.toml.
+
+Here are all the options available under it, the values set are its default.
+```toml
+[proc-macro]
+name = "macros"
+path = "src/procmacros.rs"
+test = true
+doctest = true
+bench = false
+doc = true
+proc-macro = true # (cannot be changed)
+```
+
+To disable automatic finding, use:
+```toml
+[package]
+autoprocmacro = false
+```
+
+## Implemention Details
+The package targets would be compiled in the following order:
+1. `build`
+2. `macros`
+3. `lib`
+4. `bin`s
+5. ...
+
+During compilation, it would set the `proc_macro` cfg variable (i.e. `assert!(config!(proc_macro))` would be ok in the macros crate), as well as the `CARGO_CFG_PROC_MACRO` env variable. The `OUT_DIR` environment variable would be available, with all other usually available variables.
```

Comment:
```suggestion
During compilation, it would set the `proc_macro` cfg variable (i.e. `assert!(cfg!(proc_macro))` would be ok in the macros crate), as well as the `CARGO_CFG_PROC_MACRO` env variable. The `OUT_DIR` environment variable would be available, with all other usually available variables.
```